File: #101_Symmetric_Tree/Solution2.java
/**
 * Iterative solution
 * Explicit stack use
 */

import java.util.*;

public class Solution2 {
    public boolean isSymmetric(TreeNode root) {
        if (root == null) return true;
        if (root.left == null && root.right == null) return true;
        if (root.left == null || root.right == null) return false;
        Deque<TreeNode> stack = new LinkedList();
        stack.push(root.left);
        stack.push(root.right);
        while (!stack.isEmpty()) {
        	TreeNode left = stack.pop();
        	TreeNode right = stack.pop();
        	if (left.val != right.val) return false;
        	if ((left.left == null || right.right == null) && !(left.left == null && right.right == null)) 
        		return false;
        	if (left.left != null) stack.push(left.left);
        	if (right.right != null) stack.push(right.right);
        	if ((left.right == null || right.left == null) && !(left.right == null && right.left == null)) 
        		return false;
        	if (left.right != null) stack.push(left.right);
        	if (right.left != null) stack.push(right.left);
        }
        return true;
    }

    public static void main(String[] args) {
    	Solution2 sol = new Solution2();
    	TreeNode root = new TreeNode(1);
    	root.right = new TreeNode(2);
    	System.out.println(sol.isSymmetric(root));
    }
}

File: #101_Symmetric_Tree/Solution2.py
'''
Iterative solution with explicit stack
'''

class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:
	# @param {TreeNode} root
	# @return {boolean}
	def isSymmetric(self, root):
		if not root:
			return True
		stk = [root.left, root.right]
		while stk:
			lnd = stk.pop()
			rnd = stk.pop()
			if not lnd and not rnd:
				continue
			if not lnd or not rnd:
				return False
			if lnd.val != rnd.val:
				return False
			stk.append(lnd.left)
			stk.append(rnd.right)
			stk.append(lnd.right)
			stk.append(rnd.left)
		return True